### PR TITLE
Haiku: Initial work at OS support

### PR DESCRIPTION
**Pull request by kallisti5**
These changes should be non-invasive to non-Haiku platforms. These patches were hand reworked from Neil's original Rust 1.9.0 patches. I've done some style cleanup and design updates along the way.

There are a few small additional patches to libc, rust-installer and compiler-rt that will be submitted once this one is accepted.

Haiku can be compiled on Linux, and a full gcc cross-compiler with a Haiku target is available, which means bootstrapping should be fairly easy. The patches here have already successfully bootstrapped under our haiku x86_gcc2 architecture. http://rust-on-haiku.com/wiki/PortingRust

I'll be focusing on our more modern gcc5 x86 and x86 architectures for now.

As for support, we're not seeking official support for now. We understand Haiku isn't a top-tier OS choice, however having these patches upstream greatly reduces the amount of patchwork we have to do. Mesa has Haiku code upstream, and we submit patches to keep it going. Mesa doesn't test on Haiku and we're ok with that :-)

### CODE REVIEW

File: src/libstd/sys/unix/thread.rs [hunk 1]
```diff
@@ -120,6 +120,11 @@ impl Thread {
         // Newlib, Illumos and Emscripten have no way to set a thread name.
     }
 
+    #[cfg(target_os = "haiku")]
+    pub fn set_name(_name: &CStr) {
+        // Haiku has no way to set a thread name.
+    }
+
```
Comment (Aatch) [inline src/libstd/sys/unix/thread.rs]: Can't this be merged with the above case?

Comment (avongluck-r1soft) [inline src/libstd/sys/unix/thread.rs]: Done

Comment (kallisti5) [inline src/libstd/sys/unix/thread.rs]: Done

File: src/libstd/sys/unix/fs.rs [hunk 2]
```diff
@@ -279,7 +279,12 @@ impl DirEntry {
         stat(&self.path()).map(|m| m.file_type())
     }
 
-    #[cfg(not(target_os = "solaris"))]
+    #[cfg(target_os = "haiku")]
```
Comment (alexcrichton) [inline src/libstd/sys/unix/fs.rs]: Shouldn't the haiku/solaris fallbacks here be the same?

**Comment by brson** [comment]
cc @rust-lang/libs @rust-lang/tools

**Pull request merged**